checker/src/org/checkerframework/checker/intrange/IntRangeVisitor.java

```diff
@@ -0,0 +1,205 @@
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/**
+ * Visitor for the Integer Range Type System
+ *
+ * @author JasonMrX
+ *
+ */
+public class IntRangeVisitor extends BaseTypeVisitor<IntRangeAnnotatedTypeFactory> {
+
+    private Set<Kind> coveredKinds;
+
+    private Set<TypeKind> coveredTypeKinds;
+
+    protected final AnnotationMirror EMPTYRANGE, INTRANGE, FULLINTRANGE;
+
+    public IntRangeVisitor(BaseTypeChecker checker) {
+        super(checker);
+
+        coveredKinds = new HashSet<Kind>(3);
+        coveredKinds.add(Kind.INT_LITERAL);
+        coveredKinds.add(Kind.LONG_LITERAL);
+        coveredKinds.add(Kind.CHAR_LITERAL);
+
+        coveredTypeKinds = new HashSet<TypeKind>(5);
+        coveredTypeKinds.add(TypeKind.BYTE);
+        coveredTypeKinds.add(TypeKind.SHORT);
+        coveredTypeKinds.add(TypeKind.CHAR);
+        coveredTypeKinds.add(TypeKind.INT);
+        coveredTypeKinds.add(TypeKind.LONG);
+
+        EMPTYRANGE = AnnotationUtils.fromClass(elements, EmptyRange.class);
+        INTRANGE = AnnotationUtils.fromClass(elements, IntRange.class);
+        FULLINTRANGE = AnnotationUtils.fromClass(elements, FullIntRange.class);
+    }
+
+    private boolean isCoveredKind(Kind k) {
+        return coveredKinds.contains(k);
+    }
+
+    private boolean isCoveredTypeKind(TypeKind tk) {
+        return coveredTypeKinds.contains(tk);
+    }
+
+    private long getValueFromCoveredKinds(ExpressionTree exp) {
+        switch (exp.getKind()) {
+            case INT_LITERAL:
+                return ((Number) ((LiteralTree) exp).getValue()).longValue();
+            case LONG_LITERAL:
+                return (Long) ((LiteralTree) exp).getValue();
+            case CHAR_LITERAL:
+                return (long) ((Character) ((LiteralTree) exp).getValue());
+            default:
+                throw new IllegalArgumentException(
+                        "exp should be within the covered kinds (INT_LITERAL, LONG_LITERAL, CHAR_LITERAL");
+        }
+    }
+
+    @Override
+    protected IntRangeAnnotatedTypeFactory createTypeFactory() {
+        return new IntRangeAnnotatedTypeFactory(checker);
+    }
+
+    /**
+     * Issues a warning if from &gt; to
+     */
+    @Override
+    public Void visitAnnotation(AnnotationTree node, Void p) {
+        List<? extends ExpressionTree> args = node.getArguments();
+
+        if (args.isEmpty()) {
+            // Nothing to do if there are no annotation arguments
+            return super.visitAnnotation(node, p);
+        }
+
+        Element elem = TreeInfo.symbol((JCTree) node.getAnnotationType());
+
+        if (!elem.toString().equals(IntRange.class.getName())) {
+            return super.visitAnnotation(node, p);
+        }
+
+        if (args.size() == 2
+                && args.get(0).getKind() == Kind.ASSIGNMENT
+                && args.get(1).getKind() == Kind.ASSIGNMENT) {
+
+            ExpressionTree expFrom = ((AssignmentTree) args.get(0)).getExpression();
```

Comment:
How do you know that `from` comes before `to`? Could somebody write `@IntRange(to=10, from=20)` and this code wouldn't complain?